pkg/csi/controller_server.go

```diff
@@ -0,0 +1,408 @@
+	}
+	return &csi.ControllerExpandVolumeResponse{
+		CapacityBytes:         req.CapacityRange.GetRequiredBytes(),
+		NodeExpansionRequired: false,
+	}, nil
+}
+
+func (cs *ControllerServer) validateControllerServiceRequest(c csi.ControllerServiceCapability_RPC_Type) error {
+	if c == csi.ControllerServiceCapability_RPC_UNKNOWN {
+		return nil
+	}
+
+	for _, cap := range cs.caps {
+		if c == cap.GetRpc().GetType() {
+			return nil
+		}
+	}
+	return status.Errorf(codes.InvalidArgument, "unsupported capability %s", c)
+}
+
+func (cs *ControllerServer) validateVolumeCapabilities(volumeCaps []*csi.VolumeCapability) error {
+	if volumeCaps == nil {
+		return status.Error(codes.InvalidArgument, "Volume Capabilities cannot be empty")
+	}
+
+	for _, cap := range volumeCaps {
+		if cap.GetMount() == nil && cap.GetBlock() == nil {
+			return status.Error(codes.InvalidArgument, "cannot have both mount and block access type be undefined")
+		}
+		if cap.GetMount() != nil && cap.GetBlock() != nil {
+			return status.Error(codes.InvalidArgument, "cannot have both block and mount access type")
+		}
+
+		supportedMode := false
+		for _, m := range cs.accessModes {
+			if cap.GetAccessMode().GetMode() == m.GetMode() {
+				supportedMode = true
+				break
+			}
+		}
+		if !supportedMode {
+			return status.Errorf(codes.InvalidArgument, "access mode %v is not supported", cap.GetAccessMode().Mode.String())
+		}
+	}
+
+	return nil
+}
+
+func getControllerServiceCapabilities(cl []csi.ControllerServiceCapability_RPC_Type) []*csi.ControllerServiceCapability {
+	var cscs []*csi.ControllerServiceCapability
+
+	for _, cap := range cl {
+		logrus.Infof("Enabling controller service capability: %v", cap.String())
+		cscs = append(cscs, &csi.ControllerServiceCapability{
+			Type: &csi.ControllerServiceCapability_Rpc{
+				Rpc: &csi.ControllerServiceCapability_RPC{
+					Type: cap,
+				},
+			},
+		})
+	}
+
+	return cscs
+}
+
+func getVolumeCapabilityAccessModes(vc []csi.VolumeCapability_AccessMode_Mode) []*csi.VolumeCapability_AccessMode {
+	var vca []*csi.VolumeCapability_AccessMode
+	for _, c := range vc {
+		logrus.Infof("Enabling volume access mode: %v", c.String())
+		vca = append(vca, &csi.VolumeCapability_AccessMode{Mode: c})
+	}
+	return vca
+}
+
+func (cs *ControllerServer) waitForPVCState(name string, stateDescription string,
+	predicate func(pvc *corev1.PersistentVolumeClaim) bool) bool {
+	timer := time.NewTimer(timeoutAttachDetach)
+	defer timer.Stop()
+	timeout := timer.C
+
+	ticker := time.NewTicker(tickAttachDetach)
+	defer ticker.Stop()
+	tick := ticker.C
+
+	for {
+		select {
+		case <-timeout:
+			logrus.Warnf("waitForPVCState: timeout while waiting for PVC %s state %s", name, stateDescription)
+			return false
+		case <-tick:
+			logrus.Debugf("Polling PVC %s state for %s at %s", name, stateDescription, time.Now().String())
+			existVol, err := cs.coreClient.PersistentVolumeClaim().Get(cs.namespace, name, metav1.GetOptions{})
```

Comment:
What about using `watch` interface instead of polling?

Reply:
The CSI driver is event-driven so we probably don't need to maintain cache all the time. Let's keep it simple and consistent with the longhorn implementation: https://github.com/longhorn/longhorn-manager/blob/16bb91ca66ad0badede48447a3727de3d0e71cf6/csi/controller_server.go#L712